service/follower/consensus.go

```diff
@@ -0,0 +1,242 @@
+	db       *badger.DB
+	hold     RecordHolder
+	payloads map[uint64]*Payload
+}
+
+// NewConsensus returns a new Consensus instance.
+func NewConsensus(log zerolog.Logger, db *badger.DB, hold RecordHolder) *Consensus {
+	f := Consensus{
+		log:      log,
+		db:       db,
+		hold:     hold,
+		payloads: make(map[uint64]*Payload),
+	}
+
+	return &f
+}
+
+// Root returns the root height from the underlying protocol state.
+func (c *Consensus) Root() (uint64, error) {
+	var height uint64
+	err := c.db.View(operation.RetrieveRootHeight(&height))
+	if err != nil {
+		return 0, fmt.Errorf("could not retrieve root height: %w", err)
+	}
+	return height, nil
+}
+
+// Header returns the header for the given height, if available.
+func (c *Consensus) Header(height uint64) (*flow.Header, error) {
+	c.purge(height)
+	payload, ok := c.payloads[height]
+	if !ok {
+		return nil, dps.ErrUnavailable
+	}
+	return payload.Header, nil
+}
+
+// Guarantees returns the collection guarantees for the given height, if available.
+func (c *Consensus) Guarantees(height uint64) ([]*flow.CollectionGuarantee, error) {
+	payload, ok := c.payloads[height]
+	if !ok {
+		return nil, dps.ErrUnavailable
+	}
+	return payload.Guarantees, nil
+}
+
+// Seals returns the block seals for the given height, if available.
+func (c *Consensus) Seals(height uint64) ([]*flow.Seal, error) {
+	payload, ok := c.payloads[height]
+	if !ok {
+		return nil, dps.ErrUnavailable
+	}
+	return payload.Seals, nil
+}
+
+// Commit returns the state commitment for the given height, if available.
+func (c *Consensus) Commit(height uint64) (flow.StateCommitment, error) {
+	payload, ok := c.payloads[height]
+	if !ok {
+		return flow.DummyStateCommitment, dps.ErrUnavailable
+	}
+	record, ok := c.hold.Record(payload.Header.ID())
+	if !ok {
+		return flow.DummyStateCommitment, dps.ErrUnavailable
+	}
+	return record.Commit, nil
+}
+
+func (c *Consensus) Collections(height uint64) ([]*flow.LightCollection, error) {
+	payload, ok := c.payloads[height]
+	if !ok {
+		return nil, dps.ErrUnavailable
+	}
+	record, ok := c.hold.Record(payload.Header.ID())
+	if !ok {
+		return nil, dps.ErrUnavailable
+	}
+	collections := make([]*flow.LightCollection, 0, len(record.Collections))
+	for _, complete := range record.Collections {
+		collection := complete.Collection().Light()
+		collections = append(collections, &collection)
+	}
+	return collections, nil
+}
+
+func (c *Consensus) Transactions(height uint64) ([]*flow.TransactionBody, error) {
+	payload, ok := c.payloads[height]
+	if !ok {
+		return nil, dps.ErrUnavailable
+	}
+	record, ok := c.hold.Record(payload.Header.ID())
+	if !ok {
+		return nil, dps.ErrUnavailable
+	}
+	transactions := make([]*flow.TransactionBody, 0, len(record.Collections))
+	for _, complete := range record.Collections {
+		transactions = append(transactions, complete.Transactions...)
+	}
+	return transactions, nil
+}
+
+func (c *Consensus) Results(height uint64) ([]*flow.TransactionResult, error) {
```

Comment:
Missing comment, while other methods in this file have a comment that describes the methods.